File: engine/tokenizer.js
/**
 * STEP 1 — Tokenizer (Lexer)
 *
 * Goal: Convert an equation string into tokens without evaluating it.
 * - Handles multi-digit integers
 * - Handles variables (single letters)
 * - Operators: + - * / =
 * - Brackets: ( ) [ ]
 * - Implicit multiplication: 10x -> 10 * x, 2(x+1) -> 2 * (x+1), xy -> x * y
 */

/** @typedef {{ type: 'number', value: number } | { type: 'ident', value: string } | { type: 'op', value: '+'|'-'|'*'|'/'|'=' } | { type: 'paren', value: '('|')'|'['|']' }} Token */

const OPS = new Set(["+", "-", "*", "/", "="]);
const OPEN = new Set(["(", "["]);
const CLOSE = new Set([")", "]"]);

function isWhitespace(ch) {
  return ch === " " || ch === "\t" || ch === "\n" || ch === "\r";
}

function isDigit(ch) {
  return ch >= "0" && ch <= "9";
}

function isAlpha(ch) {
  const code = ch.charCodeAt(0);
  return (code >= 65 && code <= 90) || (code >= 97 && code <= 122);
}

function tokenToString(t) {
  if (t.type === "number") return String(t.value);
  return t.value;
}

function isAtomStart(t) {
  return t.type === "number" || t.type === "ident" || (t.type === "paren" && OPEN.has(t.value));
}

function isAtomEnd(t) {
  return t.type === "number" || t.type === "ident" || (t.type === "paren" && CLOSE.has(t.value));
}

/**
 * Inserts explicit '*' where multiplication is implied.
 * Example: [10, x] -> [10, '*', x]
 *
 * Rule: If an "atom" ends and the next token is an "atom" start, insert '*'.
 * Excludes cases where either side is '='.
 *
 * Atom end: number, ident, ')', ']'
 * Atom start: number, ident, '(', '['
 */
function insertImplicitMultiplication(tokens) {
  /** @type {Token[]} */
  const out = [];

  for (let i = 0; i < tokens.length; i++) {
    const current = tokens[i];
    out.push(current);

    const next = tokens[i + 1];
    if (!next) continue;

    if (current.type === "op" && current.value === "=") continue;
    if (next.type === "op" && next.value === "=") continue;

    if (isAtomEnd(current) && isAtomStart(next)) {
      out.push({ type: "op", value: "*" });
    }
  }

  return out;
}

/**
 * Tokenize an equation string.
 * @param {string} input
 * @returns {Token[]}
 */
function tokenizeEquation(input) {
  if (typeof input !== "string") {
    throw new Error("Equation input must be a string.");
  }

  /** @type {Token[]} */
  const tokens = [];
  let i = 0;

  while (i < input.length) {
    const ch = input[i];

    if (isWhitespace(ch)) {
      i++;
      continue;
    }

    if (OPS.has(ch)) {
      tokens.push({ type: "op", value: /** @type {any} */ (ch) });
      i++;
      continue;
    }

    if (OPEN.has(ch) || CLOSE.has(ch)) {
      tokens.push({ type: "paren", value: /** @type {any} */ (ch) });
      i++;
      continue;
    }

    if (isDigit(ch)) {
      let start = i;
      while (i < input.length && isDigit(input[i])) i++;
      const raw = input.slice(start, i);
      const value = Number(raw);
      if (!Number.isSafeInteger(value)) {
        throw new Error(`Invalid or unsafe integer literal: ${raw}`);
      }
      tokens.push({ type: "number", value });
      continue;
    }

    // Variables are single letters (per project spec).
    // This makes "xy" tokenize as "x" "y" and then become "x * y" via implicit multiplication.
    if (isAlpha(ch)) {
      tokens.push({ type: "ident", value: ch });
      i++;
      continue;
    }

    throw new Error(`Unexpected character '${ch}' at position ${i}.`);
  }

  return insertImplicitMultiplication(tokens);
}

module.exports = {
  tokenizeEquation,
  insertImplicitMultiplication,
  tokenToString,
};

File: engine/rust/src/tokenizer.rs
// ============================================================
//  tokenizer.rs — Port of engine/tokenizer.js
//
//  Converts an equation string into a flat Vec<Token>.
//  Handles:
//    - integers (multi-digit)
//    - single-letter variables
//    - operators: + - * / =
//    - brackets: ( ) [ ]
//    - implicit multiplication insertion (10x → 10 * x, xy → x * y)
// ============================================================

use std::fmt;

#[derive(Debug, Clone, PartialEq)]
pub enum Token {
    Number(i64),
    Ident(char),
    Op(char),    // '+' '-' '*' '/' '='
    Paren(char), // '(' ')' '[' ']'
    UnaryMinus,  // synthetic 'u-'
}

impl fmt::Display for Token {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Token::Number(n) => write!(f, "{}", n),
            Token::Ident(c) => write!(f, "{}", c),
            Token::Op(c) => write!(f, "{}", c),
            Token::Paren(c) => write!(f, "{}", c),
            Token::UnaryMinus => write!(f, "u-"),
        }
    }
}

fn is_open_bracket(c: char) -> bool {
    c == '(' || c == '['
}

fn is_close_bracket(c: char) -> bool {
    c == ')' || c == ']'
}

fn is_op(c: char) -> bool {
    matches!(c, '+' | '-' | '*' | '/' | '=')
}

fn is_atom_end(t: &Token) -> bool {
    matches!(t, Token::Number(_) | Token::Ident(_) | Token::Paren(')') | Token::Paren(']'))
}

fn is_atom_start(t: &Token) -> bool {
    matches!(
        t,
        Token::Number(_) | Token::Ident(_) | Token::Paren('(') | Token::Paren('[')
    )
}

fn insert_implicit_mul(tokens: Vec<Token>) -> Vec<Token> {
    let mut out = Vec::with_capacity(tokens.len() * 2);

    for i in 0..tokens.len() {
        let cur = &tokens[i];
        out.push(cur.clone());

        if let Some(next) = tokens.get(i + 1) {
            // Skip if current or next is '='
            if matches!(cur, Token::Op('=')) || matches!(next, Token::Op('=')) {
                continue;
            }
            if is_atom_end(cur) && is_atom_start(next) {
                out.push(Token::Op('*'));
            }
        }
    }
    out
}

pub fn tokenize_equation(input: &str) -> Result<Vec<Token>, String> {
    let chars: Vec<char> = input.chars().collect();
    let mut tokens: Vec<Token> = Vec::new();
    let mut i = 0;

    while i < chars.len() {
        let ch = chars[i];

        if ch.is_whitespace() {
            i += 1;
            continue;
        }

        if is_op(ch) {
            tokens.push(Token::Op(ch));
            i += 1;
            continue;
        }

        if is_open_bracket(ch) || is_close_bracket(ch) {
            tokens.push(Token::Paren(ch));
            i += 1;
            continue;
        }

        if ch.is_ascii_digit() {
            let start = i;
            while i < chars.len() && chars[i].is_ascii_digit() {
                i += 1;
            }
            let raw: String = chars[start..i].iter().collect();
            let value: i64 = raw
                .parse()
                .map_err(|_| format!("Invalid integer literal: {}", raw))?;
            tokens.push(Token::Number(value));
            continue;
        }

        if ch.is_ascii_alphabetic() {
            tokens.push(Token::Ident(ch));
            i += 1;
            continue;
        }

        return Err(format!("Unexpected character '{}' at position {}", ch, i));
    }

    Ok(insert_implicit_mul(tokens))
}
